### PR TITLE
Do not throw on staged(::nonleaf)

### PR DESCRIPTION
**Pull request by carnaval**
For us poor souls in gdb with a breakpoint on jl_throw.

### CODE REVIEW

**Comment by yuyichao** [comment]
:+1: After hitting 20+ `jl_throw` breakpoint in inference this morning...

File: base/reflection.jl [hunk 1]
```diff
@@ -238,6 +238,9 @@ function code_typed(f::Function, types::ANY; optimize=true)
     asts = []
     for x in _methods(f,types,-1)
         linfo = Core.Inference.func_for_method(x[3],types,x[2])
+        if linfo === Core.Inference.NF
+            error()
```
Comment (yuyichao) [inline base/reflection.jl]: Maybe keep the error message here?

File: base/reflection.jl [hunk 2]
```diff
@@ -264,6 +267,9 @@ function return_types(f::Function, types::ANY)
     rt = []
     for x in _methods(f,types,-1)
         linfo = Core.Inference.func_for_method(x[3],types,x[2])
+        if linfo === Core.Inference.NF
+            error()
```
Comment (yuyichao) [inline base/reflection.jl]: And here.

**Comment by vtjnash** [comment]
+1999

File: base/reflection.jl [hunk 3]
```diff
@@ -263,7 +273,10 @@ function return_types(f::Function, types::ANY)
     types = to_tuple_type(types)
     rt = []
     for x in _methods(f,types,-1)
-        linfo = Core.Inference.func_for_method(x[3],types,x[2])
+        linfo = func_for_method_checked(x,types)
+        if linfo === Core.Inference.NF
```
Comment (yuyichao) [inline base/reflection.jl]: Should this branch be removed?

Comment (carnaval) [inline base/reflection.jl]: yep, although codegen should do it :p

**Pull request merged**